npm-command-guidelines.md

```diff
@@ -0,0 +1,60 @@
+ - Boilerplate project code becomes more reusable, shortening time to create new projects and encourage similarity between projects at the same time.
+
+This guide has been developed by analysing which npm commands and command patterns are already in use at Holiday Extras, and doing our best to preserve that while establishing a common standard.
+
+##Observations from our repositories
+The following observations have been made from analysing the scripts blocks of all of the `package.json` files within the Holiday Extras Github organisation's projects.
+
+###Types of command
+Our npm commands tend to be structured into 5 main categories:
+
+ - Build related commands
+ - Test related commands
+ - Run related commands
+ - CI related commands
+ - Deploy/Publish related commands
+
+There's also a long tail of misellaneous commands.
+
+###Command structure
+We seem to almost universally use `:` as a separator to namespace our commands. E.g. `build:js` and `build:css`.
+
+###Commands to run at a certain time
+We have a not insignificant number of commands with names that reflect _when_ they should be run, e.g. `prestart`, `pretest` and `postci`. This is in contrast to the majority of our commands have names that reflect _what_ the command does.
+
+##Recommendations
+Based on the above, we've come up with the following recommendations for command names.
+
+### Namespace commands
+Commands should be namespaced sensibly in a `:` separated format where possible. Common namespaces that we should aim to share between our projects should include `build`, `test`, `run`, `ci`, `deploy`, and `publish`.
+
+###Reserve the top level of a namespace for a 'do-all' command
+While we do want to namespace, we don't want to stop people from using the top-level `npm run test` or `npm run build` commands. These should, however, be reserved for a 'do everything' task _if_ one is readily available.
+
+For example, `npm run test` should run _all_ the tests of _all_ types, whereas you may have `npm run test:unit`, `npm run test:integration` and `npm run test:journey` if you wanted to split up your test types into individual commands.
+
+###Keep command names technology agnostic
+We shouldn't need to update all of our command names if we decide to switch technologies, and new developers shouldn't have to be familiar with a new technology just to get a project up and running. In support of that, we should use command names that don't reflect the underlying technology where it's not necessary:
+
+**Bad:**
+
+```
+npm run karma
+```
+
+**Good:**
+
+```
+npm run test:browser
+```
+
+###Keep commands reflecting the _what_ as well as _when_ where possible
+Commands names should succinctly reflect what the command is doing, not just when it should be run. Adding a namespace can help here. For example, instead of `npm run postci`, `npm run postci:clear-cache` might be a better command name that gives more insight as to what's going on.
```

Comment:
Will that command continue to work as before? Pre and post commands are a hook convention provided by npm. If I've got the [docs](https://docs.npmjs.com/misc/scripts) right, `postci` will run automatically after a `ci` script, but `postci:clear-cache` will run only after a `ci:clear-cache` script, which probably doesn't exist as a `ci` "sub-script". Our projects probably aren't (or probably shouldn't be) running `npm run postci` directly.